src/bot.rs

```diff
@@ -0,0 +1,175 @@
+            if let Some(game_id) = player_games.get(&player_id) {
+                if let Some(inst) = game_last_played.get_mut(game_id) {
+                    *inst = std::time::Instant::now();
+                }
+                let channel = game_channel.get(game_id).unwrap();
+                channel.send(ThreadMessage::HandleMove(primitives::Player{id: qry.from.id.into(), name: utils::get_user_name(&qry.from.first_name, &qry.from.last_name)}, card)).expect("Could not handle move");
+            } else {
+                client.send_message(("Gioco non trovato", player_id).into());
+            }
+        }
+        _ => {}
+    }
+}
+
+fn handle_update(
+    update: telegram_bot_raw::types::Update,
+    playable_games: &Vec<Box<dyn Game>>,
+    player_games: &mut HashMap<telegram_bot_raw::types::refs::UserId, String>,
+    game_channel: &mut HashMap<String, std::sync::mpsc::SyncSender<threading::ThreadMessage>>,
+    game_last_played: &mut HashMap<String, std::time::Instant>,
+    client: &mut Telegram
+) {
+    use telegram_bot_raw::types::update::UpdateKind;
+    use telegram_bot_raw::types::message::MessageKind;
+    if let UpdateKind::Message(msg) = update.kind {
+        if let MessageKind::Text{data, entities} = msg.kind {
+            drop(entities); // Silence the stupid warning and free some RAM
+            if data.contains("/start") {
+                let pieces: Vec<String> = data.split(" ").map(|x| x.to_owned()).collect();
+                if pieces.len() == 1 {
+                    client.send_message(("Ciao! A che gioco vuoi giocare?", msg.from.id, playable_games).into());
+                } else {
+                    let game_id = pieces[1].clone();
+                    add_player_to_game(game_id, client, player_games, game_channel, msg.from);
+                }
+            } else {
+                // Pass to thread
+                // It's a text message that has to be handled. If a user has more than one active game
+                // I have to ask him which one
+            }
+        } // ignoring other message kinds since they're useless for us
+    }  else if let UpdateKind::CallbackQuery(qry) = update.kind {
+        handle_callback_query(qry, playable_games, player_games, game_channel, game_last_played, client);
+    }
+}
+
+pub fn main_bot_logic(
+    playable_games: &Vec<Box<dyn Game>>,
+    player_games: &mut HashMap<telegram_bot_raw::types::refs::UserId, String>,
+    game_channel: &mut HashMap<String, std::sync::mpsc::SyncSender<threading::ThreadMessage>>,
+    game_last_played: &mut HashMap<String, std::time::Instant>,
+    client: &mut Telegram
+) {
+    let mut cleanup_list: Vec<String> = Vec::new();
+    loop {
+        for update in client.get_updates() {
+            handle_update(update, playable_games, player_games, game_channel, game_last_played, client);
+        }
+        // Now check if there are games which need to be terminated
+        for (game, time) in game_last_played.iter() {
```

Comment:
this can be a function by himself:
purge_terminated_games

Reply:
No, because it doesn't purge them

Reply:
I see you are cleaning them up later in the code
Instead of purge_terminated_games I'd call it get_terminated_games or something like that
And eventually call clean_games